### ISSUE TITLE
Chainer vs Pytorch: Re-visit computational time

### ISSUE THREAD

**Issue by kan-bayashi** [espnet/espnet#380]
We like chainer but in terms of computational time, it is much slower than pytorch.
I summarized the result of comparison between chainer and pytorch using current version.
We want developpers of chainer to accelerate the training process.

## Environments

- CentOS 7
- Python 3.6
- Cuda 9.0
- Cudnn v7.2
- NCCL v2.2
- GTX Titan pascal (x 2)
- Chainer v4.3.1
- Cupy v4.3.0
- Pytorch v0.4.1

## Experimental condition

- WSJ recipe
- ngpu = 2
- batch size 16 (resulting in 32 with the use of multi-gpu)
- wide and narrow network
   - etype = vggblstm
   - elayers = 3
   - eunits = 1024
   - atype = location
   - adim = 1024
   - dlayers = 1
   - dunits = 1024

### Training (on Multi-GPU)

| backend | iters / sec | hours / epoch | total hours |
|:-----------|:------------:|:------------:|:------------:|
| Chainer | 0.46  | 1.03 | 15.5 |
| Pytorch | 0.94  | 0.51 | 7.65 |

### Decoding (on CPU)

- Required time to decode WSJ eval data
- Joint CTC/Attention + Word LM decoding

Backend | Average [sec] | Median [sec]
:-- | :--: | :--:
Chainer (w/o MKL) | 187.31 | 171.81
Chainer (w/ MKL) | 168.56 | 151.99
Pytorch (w/o MKL) | 91.53 | 81.14
Pytorch (w/ MKL) | 105.24 | 102.75

For chainer, using MKL improve decoding speed but the effectiveness is limited because we fixed to use single thread during decoding (due to multi job submission), 
For pytorch, I did not change pytorch installation, maybe MKL did not effect the results. 

Note: Performance is almost same, CER: 3.2 (chainer) vs 3.3 (pytorch).

```python 
# w/o MKL
In [1]: np.show_config()
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
```
![](https://docs.google.com/spreadsheets/d/e/2PACX-1vQqnRjQIaReMRxhdHMHD1km_DucEYj1BXir1DOwVN0e6CalHC6rOyR8vuWom0Gd2XB0hKRY5VJwRoPW/pubchart?oid=1898623055&format=image)
![](https://docs.google.com/spreadsheets/d/e/2PACX-1vQqnRjQIaReMRxhdHMHD1km_DucEYj1BXir1DOwVN0e6CalHC6rOyR8vuWom0Gd2XB0hKRY5VJwRoPW/pubchart?oid=2021498533&format=image)

```python
# w/ MKL
In [1]: np.show_config()
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/opt/intel/mkl/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/opt/intel/mkl/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/opt/intel/mkl/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/opt/intel/mkl/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/opt/intel/mkl/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/opt/intel/mkl/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/opt/intel/mkl/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/opt/intel/mkl/include']
```
![](https://docs.google.com/spreadsheets/d/e/2PACX-1vQqnRjQIaReMRxhdHMHD1km_DucEYj1BXir1DOwVN0e6CalHC6rOyR8vuWom0Gd2XB0hKRY5VJwRoPW/pubchart?oid=1449768705&format=image)
![](https://docs.google.com/spreadsheets/d/e/2PACX-1vQqnRjQIaReMRxhdHMHD1km_DucEYj1BXir1DOwVN0e6CalHC6rOyR8vuWom0Gd2XB0hKRY5VJwRoPW/pubchart?oid=1749400150&format=image)

**Comment by kan-bayashi** [comment]
I'm not sure but in the case of chainer multi GPU, all of losses are gather into main gpu. 
Therefore, main gpu requires slightly more memory.
If the batchsize is near the limit in the case of single gpu, the problem might be happen I think.

**Comment by kan-bayashi** [comment]
Hi @ShigekiKarita.

> Instead of returning loss vectors, is it possible to summed loss scalar and number of losses (batchsize per gpu)?

In the case of pytorch, current implementation returns summed loss scarlar, I think.
But in the case of chainer, we do not modified the model itself.
We just changed `updator` to enable multi-gpus.
So I’m not sure what happened actually.

> I believe that OMP_NUM_THREADS=1 is the fastest setting (rather than OMP_NUM_THREADS=$ncpu). Do you agree?

If the server has a lot of cpus more than #jobs, $OMP_NUM_THREADS=1 is not the fastest setting, but I think this is very rare case.
In the case of usual usage, $OMP_NUM_THREADS=1 is the fastest, I think.

**Comment by ShigekiKarita** [comment]
Actually, I have many CPU cores (~50) per single job, even
OMP_NUM_THREADS=1 was fastest when I benchmarked. Maybe omp algorithms or
CPU  cores are not smart as GPU cores.

2018年9月3日(月) 16:06 Tomoki Hayashi <notifications@github.com>:
